### ISSUE TITLE
Docker client should pre-authorize token scopes before making ping

### ISSUE THREAD

**Issue by stevvooe** [moby/moby#12969]
We are getting a large number of 401s to the "/v2/" endpoint, presumably from docker clients exploring version resolution. It may be better to have the client prefetch the token to reduce the number of round trips, even if the client won't using v2.

@dmcgowan @dmp42

**Comment by dmcgowan** [comment]
This is currently part of the ping process to determine what version a registry supports as well as the authorization mechanism. Without doing this check there is no way (without hard coding a special case for the Hub) to know whether a registry expects token auth, basic auth, or no auth. Even with a discover mechanism there would be no way to guess how a registry is authorized.

**Comment by stevvooe** [comment]
@dmcgowan Do we like the current "discovery" mechanism, of having the registry tell you? Should we cache the discovered result more aggressively?

**Comment by dmcgowan** [comment]
The mechanism could be cached and only busted if there is a future 401. This would make the existing code more complicated though since a 401 after ping will result in an error being returned to the client and this sort of fallback would need to be transparent to the user. For the new v2 client we can defer the ping for auth method until a request is authorized and re-attempt authorization on 401 by busting the cache and authorizing. This would only be appropriate on requests without a body though without exposing it too high in the interface. WDYT?

**Comment by dmp42** [comment]
@dmcgowan I think something might be of.
The number of 401 /v2 outnumbers actual operations.
Let's discuss this.

**Comment by stevvooe** [comment]
@dmcgowan Is this really more complicated if we have something like a [`TokenSource`](https://godoc.org/golang.org/x/oauth2#TokenSource)? It seems like this could all be transparent.

**Comment by dmcgowan** [comment]
We can re-asses the state of the Docker code after the refactor. The current ping logic will be changing and it may make caching of the authentication scheme more practical.

How much do 401 /v2 requests outnumber actual operations, I wouldn't expect that.

**Comment by dmp42** [comment]
Example over a five minutes period:
- HTTP 200: 431 hits
- HTTP 401: 3020 hits
- HTTP 302:  788 hits

**Comment by AngryHacker** [comment]
Sorry, I don't know if it's proper to ask here. But I really confused and don't know how to solve the problem.

I want to build a private registry using registy 2.0. When I try using the auth option with silly in the config.yml, I got this error when I try to push:

FATA[0014] Error response from daemon: no successful auth challenge for http://localhost:4000/v2/ - errors: [Get http://silly-realm?account=silly&service=silly-service: dial tcp: lookup silly-realm on 192.168.11.99:53: server misbehaving] 

I know it's because the HTTP/1.1 401 Unauthorized but I don't know how to deal with it. Or how should I use the docker client to connect to a registry 2.0 which need auth.

**Comment by dmcgowan** [comment]
@AngryHacker the best place to ask that question is in the distribution repo https://github.com/docker/distribution or on irc at #docker-distribution. For a brief answer though, `silly` auth is a development setting not intended to provide authentication.

**Comment by AngryHacker** [comment]
@dmcgowan Thank you! I have got a general idea about how to solve this.

**Comment by dmcgowan** [comment]
@stevvooe this may now be possible with the refactor.  Currently the design I am planning on going with involves pinging to get the authentication challenges before creating the transport. However this ping will no longer be necessary for version detection and it should be possible to pre-set the token parameters but this will lose out on controlling the auth server location from the registry.

**Comment by stevvooe** [comment]
@tiborvass Did we actually fix this with #13375?

**Comment by tiborvass** [comment]
My bad, not sure why i added "Fixes" on my PR

**Comment by thaJeztah** [comment]
This was previously on the 1.7 milestone, but removed; should it be on the 1.9 milestone now?

**Comment by calavera** [comment]
No, unless we have a clear path to solve it and someone working on it.

**Comment by LK4D4** [comment]
@stevvooe @dmcgowan still problem? still want to fix?

**Comment by stevvooe** [comment]
@LK4D4 As far I know, it still is an issue, but we've changed the way we authorize tokens in a lot of cases to reduce problems here. I'd still like to get rid of the "ping" request, as that creates a lot of problems.

@dmcgowan may know know more, here.

**Comment by dmcgowan** [comment]
This has been something we have discussed quite a few times and I think we all agree that current ping method is not ideal and should be replaced. The alternatives to ping would involve attempting to use the registry API without a ping, and only once a 401 has been received to respond to that 401 by getting a token or adding authorization credentials and reissuing the request. In practice this will not reduce the number of 401's to registries, however it will get rid of the 401's on `/v2/` and allow the registry to response more appropriately by setting the desired scope for the request. Currently the client is responsible for setting the scope which gets sent to the authorization server. This would be nice to have but not high priority since it will not save the number of requests for most registries.

We did get rid of the version detection which for awhile was causing an extra call to `/v2/`. I agree with leaving this open as an enhancement since the original request not getting rid of the ping entirely has not been addressed.

**Comment by thaJeztah** [comment]
> The alternatives to ping would involve attempting to use the registry API without a ping, and only once a 401 has been received to respond to that 401 by getting a token or adding authorization credentials and reissuing the request. In practice this will not reduce the number of 401's to registries, however it will get rid of the 401's on /v2/ and allow the registry to response more appropriately by setting the desired scope for the request

@dmcgowan I think with the containerd image-store integration, that approach is taken? So we can probably consider this one as "done"?

**Comment by dmcgowan** [comment]
Yes, mark it as done. The ping endpoint is not used by containerd nor recommended to use anywhere. The ping was only added to differentiate between a v2 registry and the legacy v1 registry, which has long since disappeared.
